inferno-framework/inferno-core
# FI-1362: Check for missing inputs and user runnable before creating a test run

Check that test runs are have all required inputs and are user runnable _before_ persisting the test run to the database.

To test create a test run
e.g. `http://localhost:4567/inferno/api/test_sessions?test_suite_id=infra_test`

Then use the Test Session id returned try to execute a test with some inputs missing
e.g. `http://localhost:4567/inferno/api/test_runs` with a body that doesn't have all required inputs`

Previously a first attempt at running would return a

```http
HTTP 422 Unprocessable Entity
{
	"errors": "Missing the following required inputs: suite_input"
}
```

Any following test execution attempts result in the following, even if all the inputs are provided
```http
HTTP 409 Conflict
{
	"error": "Cannot run new test while another test run is in progress"
}
```

With this change any following test run attempts with all the inputs will now be successful (return an `HTTP 200` with a payload with test result details)

---

One thing I noticed while testing this is that the JSON API Swagger docs don't indicate that the input type as a required field, even though it is

![Screen Shot 2022-03-01 at 1 01 24 PM](https://user-images.githubusercontent.com/41651655/156223725-d02c03ba-993b-4fdd-97ae-1e95c65a7c79.png)

I think all we need to do is remove readOnly: true from the `swagger.yml`, but I wanted to bring this up for discussion.

https://github.com/inferno-framework/inferno-core/blob/a1c80551178226d9caa0360e1a8b3c818c25eb3f/docs/swagger.yml#L667

Should the client be required to submit input type or is this something the backend should be able to infer?

https://github.com/inferno-framework/inferno-core/blob/a1c80551178226d9caa0360e1a8b3c818c25eb3f/lib/inferno/apps/web/controllers/test_runs/create.rb#L34-L41

When types aren't provided test runs can end up in a similar limbo since the test run is created before valid input types are checked.
```http
HTTP 500 Internal Server Error

{
	"errors": "Unknown type '' for 'suite_input'."
}
```
